### ISSUE TITLE
OSX 10.12 codesign problem

### ISSUE THREAD

**Issue by gmallard** [derekparker/delve#655]
Please answer the following before submitting your issue:

Note: Please include any substantial examples (debug session output,
stacktraces, etc) as linked gists.
1. What version of Delve are you using (`dlv version`)?

```
[guyallard@hotmacma delve (master)]$ git log --oneline -1
0f4b515 proc, terminal: stepout command
[guyallard@hotmacma delve (master)]$ dlv version
Delve Debugger
Version: 0.11.0-alpha
Build: 0f4b5150c370467c6478bc73b7ed64b92d94af56
```
1. What version of Go are you using? (`go version`)?

Reasonably recent HEAD of master:

```
[guyallard@hotmacma delve (master)]$ go version
go version devel +ca4431a Fri Oct 21 23:18:10 2016 +0000 darwin/amd64
```
1. What operating system and processor architecture are you using?

```
[guyallard@hotmacma delve (master)]$ uname -a
Darwin hotmacma 16.0.0 Darwin Kernel Version 16.0.0: Mon Aug 29 17:56:20 PDT 2016; root:xnu-3789.1.32~3/RELEASE_X86_64 x86_64
[guyallard@hotmacma delve (master)]$ sw_vers
ProductName:    Mac OS X
ProductVersion: 10.12
BuildVersion:   16A323
```
1. What did you do?

Attempt to follow this instruction:

```
CERT=dlv-cert make install
```
1. What did you expect to see?

I do not know exactly , but a successful command.
1. What did you see instead?

```
[guyallard@hotmacma delve (master)]$ CERT=dlv-cert make install
go install -ldflags="-s -X main.Build=0f4b5150c370467c6478bc73b7ed64b92d94af56" github.com/derekparker/delve/cmd/dlv
codesign -s "dlv-cert"  /Users/guyallard/gosrc/bin/dlv
error: The specified item could not be found in the keychain.
make: *** [install] Error 1
```

Notes: 

I am generally trying to set this up in XCode.

The creation of the 'dlv-cert' seemed to go fine, mostly as described in instructions here. There are differences (I think) that are likely Sierra related.

I am guessing Sierra has something to do with this problem.  At this point I am sorry I upgraded my 10.11 system with no backups.

Any help / advice would be appreciated.  I prefer command line fixes if possible.

**Comment by zchee** [comment]
@gmallard I was working fine `codesign` on Sierra.
```
error: The specified item could not be found in the keychain.
```
It seems to not found `dlv-cert`.

```
security find-identity -v | grep dlv
```
Is it found in this command?
If not,

**Comment by zchee** [comment]
@gmallard I was working fine `codesign` on Sierra.
```
error: The specified item could not be found in the keychain.
```
It seems to not found `dlv-cert`.

```sh
security find-identity -v | grep dlv
```
Is it found in this command?

If not, try
https://github.com/zchee/dotfiles/blob/master/.config/codesign/codesign_setup.sh

```sh
# for dlv
./codesign_setup.sh dlv
# e.g. for lldb
./codesign_setup.sh lldb
```

Note that I use `_codesign` suffix(it's `lldb` style), so if want to use `-cert` suffix, fix
https://github.com/zchee/dotfiles/blob/master/.config/codesign/codesign_setup.sh#L10

**Comment by gmallard** [comment]
Thanks you for the effort.  Most of what you suggest does not work out of the box.   Here are many details of what I saw and tried.  The results *might* work with XCode.  I will know after a re-boot here.

This command:

```
security find-identity -v | grep dlv
```

indeed does *not* find the cert.

However this:

```
security dump-keychain
...
... // lots of removed output ...
...
keychain: "/Library/Keychains/System.keychain"
version: 256
class: 0x80001000 
attributes:
    "alis"<blob>="allard.guy.m@gmail.com"
    "cenc"<uint32>=0x00000003 
    "ctyp"<uint32>=0x00000001 
    "hpky"<blob>=0x21B001E955F7EE21E15E10B37CEB6B40C4CA3BBF  "!\260\001\351U\367\356!\341^\020\263|\353k@\304\312;\277"
    "issu"<blob>=0x3081913111300F06035504030C08646C762D6365727431143012060355040A0C0B4741204F5358204D61696E310E300C060355040B0C0547414D42503110300E06035504080C07466C6F72696461310B30090603550406130255533110300E06035504070C074F726C616E646F3125302306092A864886F70D0109011616616C6C6172642E6775792E6D40676D61696C2E636F6D  "0\201\2211\0210\017\006\003U\004\003\014\010dlv-cert1\0240\022\006\003U\004\012\014\013GA OSX Main1\0160\014\006\003U\004\013\014\005GAMBP1\0200\016\006\003U\004\010\014\007Florida1\0130\011\006\003U\004\006\023\002US1\0200\016\006\003U\004\007\014\007Orlando1%0#\006\011*\206H\206\367\015\001\011\001\026\026allard.guy.m@gmail.com"
    "labl"<blob>="dlv-cert"
    "skid"<blob>=<NULL>
    "snbr"<blob>=0x3ADFBABE  ":\337\272\276"
    "subj"<blob>=0x3081913111300F06035504030C08646C762D6365727431143012060355040A0C0B4741204F5358204D61696E310E300C060355040B0C0547414D42503110300E06035504080C07466C6F72696461310B30090603550406130255533110300E06035504070C074F726C616E646F3125302306092A864886F70D0109011616616C6C6172642E6775792E6D40676D61696C2E636F6D  "0\201\2211\0210\017\006\003U\004\003\014\010dlv-cert1\0240\022\006\003U\004\012\014\013GA OSX Main1\0160\014\006\003U\004\013\014\005GAMBP1\0200\016\006\003U\004\010\014\007Florida1\0130\011\006\003U\004\006\023\002US1\0200\016\006\003U\004\007\014\007Orlando1%0#\006\011*\206H\206\367\015\001\011\001\026\026allard.guy.m@gmail.com"
...
... // More deleted output
...
```

does.  So I got side tracked by that.

I tried that script, but could not get it to work "out of the box" for several reasons:

- Blanks at the beginning of lines
- The cat<<EOF file ends up with a literal "$CODESIGN_NAME" in the openssl config file.

I eventually mimiced the script by cut and paste.  Here is a long console log of that effort:

```

[guyallard@hotmacma delve-work (dev)]$ cat dlv_codesign_2.cfg 
[ req ]
default_bits            = 2048                  # RSA key size
encrypt_key             = no                    # Protect private key
default_md              = sha512                # MD to use
prompt                  = no                    # Prompt for DN
distinguished_name      = codesign_dn           # DN template
[ codesign_dn ]
commonName              = "dlv_codesign"
[ codesign_reqext ]
keyUsage                = critical,digitalSignature
extendedKeyUsage        = critical,codeSigning


[guyallard@hotmacma delve-work (dev)]$ /usr/local/openssl110pref/bin/openssl req -new -newkey rsa:2048 -x509 -days 3652 -nodes -config dlv_codesign_2.cfg -extensions codesign_reqext -batch -out dlv_codesign.cer -keyout dlv_codesign.key
Generating a 2048 bit RSA private key
...............+++
..............................+++
writing new private key to 'dlv_codesign.key'
-----

[guyallard@hotmacma delve-work (dev)]$ /usr/local/openssl110pref/bin/openssl x509 -in dlv_codesign.cer -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            a0:75:e8:68:d6:ac:39:85
    Signature Algorithm: sha512WithRSAEncryption
        Issuer: CN = dlv_codesign
        Validity
            Not Before: Oct 23 23:59:55 2016 GMT
            Not After : Oct 23 23:59:55 2026 GMT
        Subject: CN = dlv_codesign
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:b5:84:20:9a:60:28:a0:ac:0c:8e:26:f6:1b:e4:
                    15:c3:e1:5e:3e:64:77:ac:95:32:d8:5f:e0:0b:aa:
                    8c:03:a4:e2:59:0d:c6:fd:98:57:90:76:eb:11:8d:
                    02:50:b0:03:2e:d6:80:f2:9a:42:3b:4f:ac:8a:ac:
                    ce:a7:82:45:17:04:3d:3d:a9:2e:b8:c1:34:8a:55:
                    2d:75:a8:c2:39:58:ad:0b:77:a6:4d:62:e7:f8:64:
                    3b:33:59:32:28:4f:a3:0c:ca:89:84:f7:f0:32:94:
                    1b:66:be:df:59:f5:68:58:8d:d9:7e:87:65:7b:b1:
                    cc:90:fe:c4:3f:e4:d4:02:a4:90:25:78:47:d2:2b:
                    71:a5:81:ff:21:9e:ea:1b:bd:cd:ba:fc:78:26:70:
                    90:16:89:cb:6c:70:32:db:bb:d1:b4:38:33:b8:52:
                    c7:a4:80:97:4f:7b:5b:41:3f:44:a0:4b:29:4e:b8:
                    f2:23:f4:07:8c:61:92:fc:66:e5:79:21:52:fc:de:
                    09:e0:1c:b1:8e:b3:a4:d2:20:8f:35:60:88:04:ec:
                    06:91:15:f3:6f:d8:10:23:46:2e:11:85:e0:d7:01:
                    d0:a0:3d:6d:97:39:29:0f:a3:e4:f5:1b:af:2c:67:
                    77:04:c1:a1:bc:95:ca:b9:a5:16:1a:ea:ca:a5:cf:
                    b6:e9
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature
            X509v3 Extended Key Usage: critical
                Code Signing
    Signature Algorithm: sha512WithRSAEncryption
         2e:6d:0b:f9:5b:81:89:95:1f:b0:39:92:40:fb:f7:ce:a8:39:
         a4:0c:cc:19:29:96:2a:1a:9c:d6:0d:4a:a6:4d:fc:cd:43:bf:
         e3:2e:3c:2a:d2:df:9b:6c:e9:cb:2c:de:aa:09:62:d8:75:0e:
         18:59:5d:f9:e4:50:27:eb:eb:4c:dd:1b:24:6b:5a:4b:74:62:
         75:fc:08:12:93:cf:8d:54:9f:9c:c7:50:e8:2f:fa:4d:4c:2d:
         79:1f:13:1b:0a:2d:d0:01:25:7c:53:ee:a3:0e:e1:2f:fd:9e:
         a9:3a:87:e6:19:52:88:11:99:79:2c:ca:77:b2:0d:f4:63:66:
         ac:79:6f:7f:0d:30:a7:31:c3:24:37:d9:d3:15:c6:56:69:2e:
         e2:a7:60:7a:4a:03:24:d7:1e:c4:38:4b:80:06:fd:42:99:ef:
         13:c7:be:bd:58:b5:05:57:e1:66:9e:16:15:b5:1d:ad:57:60:
         2f:c0:2b:cc:00:0e:2d:92:84:b4:ab:76:98:3f:c1:58:30:5b:
         a6:46:a4:a8:67:d5:b4:6a:8d:ab:6f:39:62:f9:df:a3:b5:80:
         0d:1b:8f:76:d5:7a:ba:ad:b4:8b:26:54:e7:96:5e:cc:55:c1:
         5b:a4:3a:c7:27:9f:41:70:64:e2:b1:2f:bf:31:6d:84:5c:47:
         0f:84:0d:af
[guyallard@hotmacma delve-work (dev)]$ sudo security add-trusted-cert -d -r trustRoot -k /Library/Keychains/System.keychain dlv_codesign.cer
Password:

[guyallard@hotmacma delve-work (dev)]$ sudo security import dlv_codesign.key -A -k /Library/Keychains/System.keychain
Password:
1 key imported.

[guyallard@hotmacma delve-work (dev)]$ # A sanity check
[guyallard@hotmacma delve-work (dev)]$ security find-identity -v | grep dlv
  2) 10E65BE059BCB30A73528EE7B8D26C1A688AA2AC "dlv_codesign"

###
### Well, OK :-) That looks promising .....
###

[guyallard@hotmacma delve-work (dev)]$ # codesigning
[guyallard@hotmacma delve-work (dev)]$ codesign -s dlv_codesign $(which dlv)

[guyallard@hotmacma delve-work (dev)]$ # need once
[guyallard@hotmacma delve-work (dev)]$ sudo security authorizationdb write system.privilege.taskport allow
YES (0)

[guyallard@hotmacma delve-work (dev)]$ # need once
[guyallard@hotmacma delve-work (dev)]$ DevToolsSecurity -enable
Developer mode is already enabled.

###############################################################################
### I am rebooting.  We will see how XCode does with this ......
###############################################################################
```

I will let you know how this goes with XCode later.

Thanks again for the help / ideas.

**Comment by aarzilli** [comment]
You could try either using the [homebrew recipe](https://github.com/go-delve/homebrew-delve) or reading it to see what command it executes. I must also mention that there seems to be some unsolved problem with delve under mac os x 10.12: https://github.com/derekparker/delve/issues/645

**Comment by gmallard** [comment]
Several comments ......

Y'all can close this if you would like.  I managed to get where I wanted to be.  I do think you might add some documentation saying 'look out for this on OSX 10.12'.

I think I am also seeing something similar to #645.  The delve app is not the only code that seems to not work right in 10.12, see below.

I won't try that brew recipe ...... because I do not run brew. I generally build something if I need it.

A short rant follows.  There are a number of annoying and/or totally wrong things with 10.12 IMO.

- git builds, but fails at runtime in odd ways - usually silently which is deadly.
- openssl builds, but has the same characteristics as the git builds.
- go builds.  After a lot of experimentation.  And seems to run OK so far.
- Ruby builds and seems to run OK (using a custom OpenSSL environment of course.  Apple broke SSL somewhere in 10.8 IIRC).
- not to mention the total lockout of updating /usr/bin because if SIP.  I have stale items there that I cannot rm.  This is a showstopper for me at least.  I won't ever buy another Mac unless this is backed out.
- and the partial lockout of updating /etc.  I have some stale stuff in /etc/profile.  And now I cannot get rid of it.
- I build four different Java/Scala based message brokers.  The maven builds seem to work (as well as maven ever does anyway).  And the resulting broker builds all seem to run OK.
- hg builds and runs.  That is mostly python so perhaps this is not surprising.
- I would try a build of gcc/g++ in order to get a good compiler.  That is not too difficult, just time consuming.  But I would end up with a compiler I do not trust, because I do not trust the clang/gcc tools used to run the build.

The bottom line is that I would advise any one who asks me,  that an upgrade to 10.12 should be approached very carefully.

 Thanks again.

Regards, Guy
